File: leetcode/problems/valid_parentheses.rb
def is_valid(s)
  valid = {
    ')' => '(',
    '}' => '{',
    ']' => '['
  }

  seen = []

  s.each_char do |c|
    if valid.key?(c)
      return false if seen.empty? || seen.pop != valid[c]
    else
      seen.push(c)
    end
  end

  seen.empty? ? true : false
end

File: leetcode/javascript_problems/valid_parentheses.js
// Given a string containing just the characters '(', ')', '{', '}', '['
// and ']', determine if the input string is valid.
//
// The brackets must close in the correct order, "()" and "()[]{}" are
// all valid but "(]" and "([)]" are not.

var isValid = (s) => {
  let parentheses = {
    ')' : '(',
    ']' : '[',
    '}' : '{'
  };

  let chars = [];

  for (let i = 0; i < s.length; i++) {
    let c = s[i];
    if (parentheses[c]) {
      if (chars.length === 0 || chars.pop() !== parentheses[c]) {
        return false;
      }
    } else {
      chars.push(c);
    }
  }

  return chars.length === 0;
};
